cardano-crypto-class/src/Cardano/Crypto/PackedBytes.hs

```diff
@@ -0,0 +1,177 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+module Cardano.Crypto.PackedBytes (PackedBytes(..), packBytes, unpackBytes) where
+
+import Control.DeepSeq
+import Data.ByteString.Short
+import Data.ByteString.Short.Internal
+import Data.Primitive.ByteArray
+import Data.Typeable
+import Data.Word
+import GHC.Exts
+import GHC.ST
+import GHC.TypeLits
+import NoThunks.Class
+
+data PackedBytes (n :: Nat) where
+  PackedBytes28# :: Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> Word16#
+                 -> PackedBytes 28
+  PackedBytes32 :: {-# UNPACK #-} !Word64
+                -> {-# UNPACK #-} !Word64
+                -> {-# UNPACK #-} !Word64
+                -> {-# UNPACK #-} !Word64
+                -> PackedBytes 32
+  PackedBytes# :: ByteArray# -> PackedBytes n
```

Comment:
Is there constructor overhead here?

Reply:
There is no pointer overhead, because of a GADT  it is statically known at compile time which constructor is being used. However, because of the fact that GADTs can't be unpacked there will be an extra word overhead for the pointer, eg:

```haskell
data Ix28 = Ix28 !(PackedBytes 28) {-# UNPACK #-} !Word64
```

Reply:
Cool, wasn't aware of that :-)